(PUP-9457) Add method for creating HTTPS connections from an SSLContext

joshcooper opened the pull request:
~PR https://github.com/puppetlabs/puppet/pull/7394 must be merged first (edit: done).~

This PR introduces a new method `Puppet::Network::HttpPool.connection` for creating HTTP(S) connections, for example:

    http = Puppet::Network::HttpPool.connection(URI('http://www.google.com'))
    puts http.get('/')
    https = Puppet::Network::HttpPool.connection(URI("https://puppet:8140"), ssl_context: ctx)
    puts http.get('/status/v1/simple/master')

The `ssl_context` option is required when using an "https" scheme.

This PR introduces a `Puppet::SSL::Verifier` class to supersede `Puppet::SSL::Validator*`, which has a number of issues:

 * The default validator will lazily lookup `Puppet::SSL::Host` to get the current host's cert and key.
 * That can cause additional cert related network connections, eg download the CRL
 * The validator gets confused about VERIFY_PEER
 * The `Puppet::SSL::Host#localhost` method will cause a settings catalog to be applied.
 * The default validator's setup_connection method is called for non-SSL connections, even though the validator is completely unnecessary.

The `Verifier` has methods for setting up the connection and handling SSL errors. In the process, it fixes the certname/hostname mismatch, (see PUP-8213).

This PR deprecates the other `HttpPool.http_instance` and `HttpPool.http_ssl_instance` methods, but preserves their behavior using a `ValidatorAdapter` to wrap the deprecated `Validator`.

puppetcla commented:
CLA signed by all contributors.

joshcooper commented on lib/puppet/util/ssl.rb:
```diff
@@ -65,26 +65,22 @@ def self.is_possibly_valid_dn?(dn)
+    # If cert/hostname are mismatched, openssl 1.1+ will generate a
+    # `verify_error` whose error code is `V_OK`, since nothing is wrong with the
+    # cert chain itself. So check for mismatch ourselves using the same routine
+    # that ruby's `post_connection_check` would do (and does with openssl 1.0)
```
Turns out the issue is not openssl 1.1, but rather ruby/openssl introduced a new `:verify_hostname` option, which defaults to true. When enabled, ruby will automatically perform hostname verification. This is generally a good thing, since applications don't need to remember to call `post_connection_check`, but in our case, it breaks our cert-mismatch detection. Need to rework the test and update these comments.

Magisus commented:
It's nice to see your previous PR being used in context. In general this definitely looks like it makes a lot more sense than what we had before, a lot more control over connection state.

joshcooper commented:
- [ ] create separate PR for verify_hostname issue against 5.5.x
- [ ] update this PR based on above
- [ ] don't reuse connections with different ssl contexts

justinstoller commented on lib/puppet/ssl/verifier.rb:
```diff
@@ -0,0 +1,75 @@
+    http.verify_callback = self
+  end
+
+  # This method is called if `HTTP::Net#start` raises an exception, which
```
`Net::HTTP`?

justinstoller commented on lib/puppet/ssl/verifier.rb:
```diff
@@ -0,0 +1,75 @@
+    @errors = []
+  end
+
+  def setup_connection(http)
```
I assume the argument is a `Net::HTTP`(`S`) instance?

joshcooper commented:
Blocked on https://github.com/puppetlabs/puppet/pull/7394

joshcooper commented on lib/puppet/ssl/verifier.rb:
```diff
@@ -16,6 +18,17 @@ def initialize(ssl_context)
+  # @param verifier [Puppet::SSL::Verifier] the verifier to compare against
+  # @return [Boolean] return true if a cached connection can be used, false otherwise
+  def reusable?(verifier)
+    verifier.is_a?(self.class) &&
```
Want verifier.class == self.class, not sure if `is_a?` will match a subclass?

joshcooper replied:
Changed to `instance_of?` which ensures it's an instance of specific class

kris-bosland commented:
Jenkins please test this on windows2012r2-64a

melissa commented on lib/puppet/network/http/connection.rb:
```diff
@@ -56,7 +56,7 @@ def initialize(host, port, options = {})
       options = OPTION_DEFAULTS.merge(options)
       @use_ssl = options[:use_ssl]
-      @verify = options[:verify]
+      @verifier = Puppet::SSL::VerifierAdapter.new(options[:verify])
```
what happens if `options[:verify]` is already the new `verifier` rather than the old `verify`?

melissa replied:
that's not great wording.

what if `options[:verify]` is already a `Verifier` rather than the old `Puppet::SSL::Validator`?

joshcooper replied:
I think it's unlikely, because callers don't call `Connection.new` directly. Instead they create a connection from the `HttpPool` factory whose methods supply the correct verifier or validator.

joshcooper replied:
Given the confusion around `verify` vs `verifier` vs `validator` I could imagine someone passing in a `Validator` when a `Verifier` is expected or vice-versa. Part of the problem is the [`DefaultValidator` doesn't extend `Validator`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/ssl/validator/default_validator.rb#L9) so I can't just check for `is_a?(Puppet::SSL::Validator)`. Also the `http_ssl_instance` method allows arbitrary validators to be passed in. But I could check for the case where a `Verifier` is passed in, and make sure that's used correctly.

joshcooper replied:
Updated

melissa commented on lib/puppet/network/http/pool.rb:
```diff
@@ -71,7 +71,8 @@ def close_connection(site, http)
   def borrow(site, verifier)
     @pool[site] = active_sessions(site)
-    session = @pool[site].shift
+    index = @pool[site].index { |session| verifier.reusable?(session.verifier) }
```
To make sure I'm understanding this, this will return the index of the first session it finds that is reusable?

joshcooper replied:
Yeah, AFAIK ruby doesn't have an enumerable method to "delete the first element for which the block is true". The `Enumerable#delete_if` method unfortunately deletes **all** elements for which the block is true.

joshcooper commented on lib/puppet/ssl/verifier.rb:
```diff
@@ -0,0 +1,132 @@
+      # if `SSLSocket#post_connection_check` raises (ruby < 2.4), then peer_cert will be valid
+      raise cert_mismatch_error(peer_cert, host)
+    elsif !@errors.empty?
+      err = @errors.first
```
When the agent verifies the server's cert chain, ruby/openssl will call our `call` method below, at least once per cert. When running ruby >= 2.4, we need to check if one of the verification errors was due to a cert/hostname mismatch. I could check each verification error to see if it's a cert mismatch, but that seems not super useful. So I just check the first error, and either raise the `CertMismatchError` or raise the first error itself (which will be a `CertVerifyError`).

melissa replied:
So if the first error isn't a cert mismatch, we don't care if any other errors are a cert mismatch?

joshcooper replied:
The case where `@errors.length > 1` and there's a cert mismatch, is if there's something wrong with one of the CA certs in the chain and the server's cert doesn't match. Like if the agent tried to send a report to a third-party SSL server, and it doesn't trust the CA cert:

```puppet
$ bx puppet agent -t --report_server https://<wronghostname>`
```

In that case, we'll report on the first verification error will be the untrusted CA cert (since the chain is processed from root to leaf), and we'll raise that. It kind of doesn't matter that there was a later cert error, until the first issue is resolved.

joshcooper replied:
ohhhh, actually the `Verifier` returns false as soon as it gets an error, so there will only be one error

joshcooper commented on lib/puppet/ssl/verifier.rb:
```diff
@@ -0,0 +1,132 @@
+  #
+  # @api private
+  def call(preverify_ok, store_context)
+    #puts "VERIFY ok=#{preverify_ok}, err=#{store_context.error} string=#{store_context.error_string}"
```
I'll take it out

joshcooper replied:
Done

joshcooper commented on lib/puppet/ssl/verifier.rb:
```diff
@@ -0,0 +1,132 @@
+      err = Puppet::SSL::CertVerifyError.new(store_context.error_string, store_context.error, peer_cert)
+
+      case store_context.error
+      when OpenSSL::X509::V_ERR_CRL_NOT_YET_VALID
```
This comes from the existing `DefaultValidator` behavior (I wanted the new Verifier to have feature parity): https://github.com/puppetlabs/puppet/blob/master/lib/puppet/ssl/validator/default_validator.rb#L81-L97 I added a case because I anticipate adding additional cases in the future. For example, it's a bit odd that we do this for CRLs but not certs.

joshcooper commented on lib/puppet/ssl/verifier.rb:
```diff
@@ -0,0 +1,132 @@
+      when OpenSSL::X509::V_ERR_CRL_NOT_YET_VALID
+        crl = store_context.current_crl
+        if crl && crl.last_update && crl.last_update < Time.now + FIVE_MINUTES_AS_SECONDS
+          Puppet.debug("Ignoring CRL not yet valid, current time #{Time.now.utc}, CRL last updated #{crl.last_update.utc}")
```
It makes total sense when I say it in a yoda voice! How about "You must unlearn what you have learned CRL" 

I actually copied that from https://github.com/puppetlabs/puppet/blob/master/lib/puppet/ssl/validator/default_validator.rb#L87

joshcooper replied:
Since the current wording matches what's currently there, I'm going to resolve this, but I'd be completely open to changing the wording to something else in a future PR.

>

joshcooper commented on lib/puppet/network/http_pool.rb:
```diff
@@ -28,6 +28,7 @@ def self.http_client_class=(klass)
   # @param verify_peer [Boolean] Whether to verify the peer credentials, if possible. Verification will not take place if the CA certificate is missing.
   # @return [Puppet::Network::HTTP::Connection]
   #
+  # @deprecated Use {#http_connection} instead.
```
There is a `Puppet.deprecation_warning` which will log once per line of code, but I don't want to enable it right now, since everything in puppet is still using the deprecated methods. I think it's best to wait until after the 6.4 release goes out, and emit the deprecation in a future release, but well before 7. I'll add a ticket to do that.

joshcooper replied:
Filed as https://tickets.puppetlabs.com/browse/PUP-9513

joshcooper commented on spec/unit/network/http_pool_spec.rb:
```diff
@@ -46,6 +46,29 @@ def initialize(host, port, options = {})
+      end
+
+      it 'does not use SSL for schemes other than HTTPS' do
+        expect(Puppet::Network::HttpPool.connection(URI('http://me'))).to_not be_use_ssl
```
Yeah probably should restrict the acceptable schemes to just `http/https`. I wouldn't expect `file://` to work in any way. I'll add a test/commit for that.

joshcooper commented on lib/puppet/network/http_pool.rb:
```diff
@@ -51,11 +52,32 @@ def self.http_instance(host, port, use_ssl = true, verify_peer = true)
+
+      verifier = Puppet::SSL::Verifier.new(ssl_context)
+      http_client_class.new(uri.host, uri.port, use_ssl: true, verifier: verifier)
+    else
```
Probably should be a warning that it will be ignored.

joshcooper commented on lib/puppet/ssl/verifier.rb:
```diff
@@ -0,0 +1,130 @@
+
+    unless preverify_ok
+      peer_cert = store_context.current_cert
+      err = Puppet::SSL::CertVerifyError.new(store_context.error_string, store_context.error, peer_cert)
```
Probably clearer to move it below, when we know we're going to add it to `@errors`

joshcooper commented:
@melissa @kris-bosland I think I've addressed your feedback in https://github.com/puppetlabs/puppet/compare/f8c869e39a8b8303dcb4aebaeb3d5047ef7d1611..c18729b421cb51a6871edabd098801865823bdc8 and https://github.com/puppetlabs/puppet/compare/c18729b421cb51a6871edabd098801865823bdc8..6f16229565c605df203e4da2c5fe6e928fb4bd3f

kris-bosland commented:
I'm 👍

joshcooper commented:
Per discussion, should only capture a single error, and report on that. Need to look at PUP-8213

joshcooper commented on lib/puppet/network/http_pool.rb:
```diff
@@ -51,11 +52,37 @@ def self.http_instance(host, port, use_ssl = true, verify_peer = true)
+      raise ArgumentError, "An ssl_context is required for HTTPS connections: #{uri}" unless ssl_context
+
+      verifier = Puppet::SSL::Verifier.new(ssl_context)
+      http_client_class.new(uri.host, uri.port, use_ssl: true, verifier: verifier)
```
Push Verifier into constructor for Connection

joshcooper commented:
- [ ] increase connect timeout
- [ ] use `include_examples`

jhelwig commented on spec/unit/ssl/verifier_spec.rb:
```diff
@@ -0,0 +1,124 @@
+#! /usr/bin/env ruby
```
Can this spec actually be run as a script? Nearly all of our spec files can't, so I've been pulling out the shebang line to people don't try, and get confused about why it's not working.

joshcooper replied:
Ah thanks! copy/pasta mistake, let me remove that.

joshcooper commented on lib/puppet/network/http/connection.rb:
```diff
@@ -56,7 +57,17 @@ def initialize(host, port, options = {})
+
+          @verifier = options[:verifier]
+        else
+          @verifier = Puppet::SSL::VerifierAdapter.new(options[:verify])
```
Previously, if `@use_ssl=true`, but `options[:verify]=false`, then the connection would raise `NoMethodError` when trying to call `verify.setup_connection(http)` for both the caching and non-caching pools.

With this change, we'll create a `VerifierAdapter` around the `nil` verify, and the connection will raise `NoMethodError` when the [adapter's `setup_connection` method is called](https://github.com/puppetlabs/puppet/pull/7397/files/32bbccc752c712a9883de2a693cfa79d0c91053d#diff-7090b16edf3f2f3b76bd2779c82b48deR33)

melissa replied:
```
  # Configure the `http` connection based on the current `ssl_context`.
  #
  # @param http [Net::HTTP] connection
  # @api private
  def setup_connection(http)
    @validator.setup_connection(http)
  end
```
Will just error with something like, can't call `setup_connection` on a nil object?

Cool, that makes a lot of sense. Thank you for clarifying that for me

kris-bosland commented on spec/lib/puppet_spec/https.rb:
```diff
@@ -0,0 +1,166 @@
+    errors
+  end
+
+CA_CERT = <<END
```
Is it reasonable to pass these artifacts in on initialize so that they can come from fixtures?  They need to match the ones used by the client, right?

joshcooper replied:
> They need to match the ones used by the client, right?

It's more the other way around. The server can use whatever certs/keys it wants, but in order for the client to authenticate the server, then the client needs to trust the server's CA cert "out of band". But the CA cert is the only shared state between the two.

The client may not have a client cert, or it can use the server's cert, which simulates an agent running on the puppetserver's host.

kris-bosland replied:
So should both this server, and the client being tested, use CA files from [here](https://github.com/puppetlabs/puppet/blob/3d046400a87ff1afe10fdd511e3530ff19766d5e/tasks/generate_cert_fixtures.rake#L48)?
